simulator/ssh/client.go

```diff
@@ -0,0 +1,211 @@
+// Package ssh provides just enough SSH/SFTP functionality to perform random writes
+// to an SFTP server.
+package ssh
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"time"
+
+	sdipacket "github.com/alphasoc/sftp-data-input/packet"
+	sdifxp "github.com/alphasoc/sftp-data-input/packet/fxp"
+
+	"golang.org/x/crypto/ssh"
+)
+
+// SSH/SFTP client version.  We don't perform any version negotiation.
+const ClientVer = 3
+
+// Only SSH_FX_OK error code is needed for our purposes.
+const SSH_FX_OK = 0
+
+// SSH/SFTP requests require an ID.  Per
+// https://datatracker.ietf.org/doc/html/draft-ietf-secsh-filexfer-13#section-4, they
+// don't need to be unique.  Since we're not sending requests/data in parallel, an id of
+// 1 is just fine.
+const reqID = 1
+
+// Client wraps SSH client and session structs.
+type Client struct {
+	sshClient *ssh.Client
+	sess      *ssh.Session
+	w         io.WriteCloser
+	r         io.Reader
+}
+
+// NewClient initializes and returns a Client ready to be used for SSH/SFTP transfer along
+// with an error.  Note that Teardown should be called when the Client is no longer needed.
+func NewClient(user string, src net.IP, dst string, signer ssh.Signer) (*Client, error) {
+	// ClientConfig will use pubkey auth, ignore the host key and apply a 5 second connection
+	// timeout.
+	config := &ssh.ClientConfig{
+		User: user,
+		Auth: []ssh.AuthMethod{
+			ssh.PublicKeys(signer),
+		},
+		// TODO insecure ok?
+		//HostKeyCallback: ssh.FixedHostKey(hostKey),
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+		Timeout:         5 * time.Second,
+	}
+	// Create dialer with src address, dial and setup a new client.
+	d := &net.Dialer{
+		LocalAddr: &net.TCPAddr{IP: src},
+		Timeout:   config.Timeout,
+	}
+	nConn, err := d.Dial("tcp", dst)
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect: %v", err)
```

Comment:
As per https://pkg.go.dev/fmt#Errorf, we should use `%w` everywhere where we pass `err`. This is quite new thing so maybe we should also update rest of the code.

Reply:
:+1:

I'll save updating the rest for another PR.